### PR TITLE
Rcd: disposable shared file link of rclone server

### PR DESCRIPTION
**Pull request by ElonH**
#### What is the purpose of this change?

implement #4352

#### Was the change discussed in an issue or in the forum before?

https://forum.rclone.org/t/rcd-disposable-shared-file-link-of-rclone-server/17090

#### Checklist

-   [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
-   [x] I have added tests for all changes in this PR if appropriate.
-   [ ] I have added documentation for the changes if appropriate.
-   [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
-   [ ] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: fs/rc/rcserver/rcserver.go [hunk 1]
```diff
@@ -188,6 +189,13 @@ func (s *Server) handler(w http.ResponseWriter, r *http.Request) {
 	}
 	path := strings.TrimLeft(urlPath, "/")
 
+	// Serve /share/links/${token}/${shared name} files
+	// TODO: need Access-Control-Allow-Origin ?
```
Comment (ElonH) [inline fs/rc/rcserver/rcserver.go:193]: I confused if `Access-Control-Allow-Origin` is needed in shared links.

File: fs/rc/rcserver/rcserver_test.go [hunk 2]
```diff
@@ -539,6 +539,151 @@ func makeMetricsTestCases(stats *accounting.StatsInfo) (tests []testRun) {
...
+			}`,
+			ContentType: "application/json",
+			Status:      http.StatusOK,
+			Expected: `{
+	"sharedLink": "share/links/123456/file.txt"
+}
+`,
+		}, {
+			Name:     "share-get-content",
+			URL:      "share/links/123456/file.txt", // TODO: it can be fetched without Authentication Header, btw can't test that here
```
Comment (ElonH) [inline fs/rc/rcserver/rcserver_test.go:612]: I don't know how to set up the test for fetching data without  Authentication Header

Comment (ivandeex) [reply fs/rc/rcserver/rcserver_test.go:612]: @ncw @negative0  Are there examples in the integration suite?

**Comment by ivandeex** [comment]
@ncw @negative0 Are you aware of any examples of a test for fetching data without Authentication Header in the integration suite?

**Comment by ncw** [comment]
You can turn the auth on and off by setting a different `NoAuth` parameter.

Here is an example

https://github.com/rclone/rclone/blob/bae550c71e80c3dc6afdf012c18c6d634ef10170/fs/rc/rcserver/rcserver_test.go#L714

**Comment by ElonH** [comment]
@ncw That is irrelevant with this feature.

If rclone serve without AuthorizationHeader, this feature is meanless.

The shared links are used in case which rclone serve is configured with Authorization. Everything need Authentication Header except these shared links `share/links/xxxxx`

**Comment by ivandeex** [comment]
@ElonH @ncw

If we add a per-request header override like `request.Headers = map{"YourAuthorizationHeaderName": ""}`, would it remove it?

**Comment by ivandeex** [comment]
@ElonH 

It turns out that rc server test in fact bypassed network stack, called internal handler directly
and consequently bypassed authentication.
I rebased your work on the latest master and patched test harness to get what you wanted.
Please pull the final code from github.

You can now proceed with your work.
Let us know when you are ready for review (post a comment below).

File: fs/rc/rcserver/rcserver_test.go (hunk 2 above)
Comment (ivandeex) [reply fs/rc/rcserver/rcserver_test.go]: fixed

**Comment by ivandeex** [comment]
@ElonH 

I finally found time to look deeper in your test code.

It turns out that rc-server test in fact bypassed network stack, called internal handler directly
and consequently authentication was not really involved in tests, just mocked.

I added a commit that really spawns a test rc-server on a random local port and uses real http client to send test requests to it, instead of mocking.
This allows to more realistically test anonymous vs authenticated access right as you wanted.

Additionally, I rebased your work on the latest master.
Please pull the final code from github (don't force-push now!)

You can proceed with your work.
Let us know when you are ready for review!
Great stuff, by the way :-)

**Comment by ivandeex** [comment]
@ElonH 
Where are you with this PR?
What do you think is missing? Can it be reviewed?

**Comment by ElonH** [comment]
@ivandeex 
Oh, sorry! I forgot this PR.
everything is fine.

**Comment by ivandeex** [comment]
@ncw 
Please take a look when you have time

**Comment by ivandeex** [comment]
@ncw I think it's a good time to look at this PR.

**Comment by ncw** [comment]
@ElonH I've had a look through the code - there are some little things that need changing if we merge it, but I thought I'd start with the bigger questions...

- What is the use case for this code?
- What are you trying to achieve with it?
- Is it in use at the moment?
- Do we want feature in rclone?

It is a nice feature, that is for sure, but it is quite complicated and doesn't persist the shared tokens to disk, meaning that if rclone is restarted the shared tokens are gone which limits its usefulness I think. So to be properly useful it needs to be more complicated. It also requires the RC interface to be exposed to the users which isn't a good design.

I'm worried that the feature is too niche for rclone.

So how else could it be implemented?

You could implement this with the [auth proxy](https://rclone.org/commands/rclone_serve_webdav/#auth-proxy) for `rclone serve` - this passes a user/pass (or token) to your program which then returns a backend or file to serve to the user. This doesn't work with `rclone serve http` yet, but it could be made to. The tokens could be stored in a directory on disk. You could even create them via the rc by writing to that local directory. So I think you could re-implement this with a little bit of coding and using the auth-proxy interface.

Thoughts?

**Comment by ivandeex** [comment]
@ncw
I can guess from the radio silence that @ElonH does not have enough time or energy to completely overhaul their work to your requirements.

**Comment by ElonH** [comment]
@ivandeex @ncw 
I owned NAS device, and rarely use rclone now.
This feature has no very useful for me.
Does someone need this feature?
I can close this PR if no one want this.

**Comment by ivandeex** [comment]
This PR is abandoned and waiting for volunteers.
cc @ncw

**Comment by ivandeex** [comment]
Closing as abandoned

**Pull request closed without merging**